keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,806 @@
+*   When pods request for a specific [RunTimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/) (`spec.runtimeClassName)`, we currently use node selectors to make sure they are on appropriate nodes . Instead, kube-scheduler can automatically identify nodes that possess the corresponding capability.
+
+## Design Details
+
+### API Changes
+
+ \
+Add a field `NodeCapabilities` field as type `map[string]string `to the` NodeSpec.NodeStatus` structure. 
+
+```
+type NodeStatus struct {
+    // ... existing fields
+    // NodeCapabilities provides a structured way to report various capabilities of the node. Keys are typically DNS-style and values are strings.
+    // +optional 
+    // +featureGate=NodeCapabilities
+    NodeCapabilities map[string]string `json:"capabilities,omitempty"`
+}
+
+
+// Node object remains unchanged in spec, only status is modified.
+type Node struct {
+    ...
+    Status NodeStatus `json:"status,omitempty" `
+}
+
+```
+
+**Example:**
+
+```
+NodeCapabilities:
+   // Kubelet configurations
+   kubernetes.io/feature/podLevelResources: "true"
+   kubernetes.io/feature/inPlacePodResize: "true"
+   // Runtime configurations 
+   runtime.kubernetes.io/handlers: "runc, runsc"
+   // Kernel capabilities
+   kernel.kubernetes.io/LSMConfig: "seLinux"   
+   // Plugins capabilities
+   csi.kubernetes.io/my-driver/feature/encryptionSupport: "true"
+```
+
+**Note:**  
+
+*   We currently have[ Node Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6279) and[ Node Runtime Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6251) which publish some runtime features through Node Status. They are too narrowly scoped and currently not used for scheduling pods. We can deprecate those fields and introduce them into NodeCapabilities if required. 
+*   Other API options are explored in the [Alternatives](?tab=t.0#heading=h.vfsx3j6ygclj) section. \
+
+
+### Kube-Scheduler Changes
+
+To enable the scheduler to filter nodes based on advertised capabilities, this KEP proposes the introduction of a new [Filter Plugin](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#filter) - `NodeCapabilityFilter`. This plugin will be part of the default scheduler and will be enabled when the `NodeCapabilities` feature gate is active. Additionally, the filter plugin would skip any capability checks for daemonset pods.
+
+`NodeCapabilityFilter` plugin would 
+
+*   Inspect the PodSpec to infer a set of required NodeCapability key-value pairs.
```

Comment:
this inference is confusing to me. How will the scheduler know which features in a pod map to one on a node? is it going to be declarative within kubernetes, or imperative?